src/test/validation_block_tests.cpp

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <iomanip>
+#include <memory>
+#include <sstream>
+
+#include <boost/test/unit_test.hpp>
+
+#include <chainparams.h>
+#include <consensus/merkle.h>
+#include <consensus/validation.h>
+#include <miner.h>
+#include <random.h>
+#include <test/test_bitcoin.h>
+#include <validation.h>
+#include <validationinterface.h>
+
+struct RegtestingSetup : public TestingSetup {
+    RegtestingSetup() : TestingSetup(CBaseChainParams::REGTEST) {}
+};
+
+BOOST_FIXTURE_TEST_SUITE(validation_block_tests, RegtestingSetup)
+
+struct TestSubscriber : public CValidationInterface {
+    uint256 m_expected_tip;
+
+    TestSubscriber(uint256 tip) : m_expected_tip(tip) {}
+
+    void UpdatedBlockTip(const CBlockIndex* pindexNew, const CBlockIndex* pindexFork, bool fInitialDownload)
+    {
+        BOOST_CHECK_EQUAL(m_expected_tip, pindexNew->GetBlockHash());
+    }
+
+    void BlockConnected(const std::shared_ptr<const CBlock>& block, const CBlockIndex* pindex, const std::vector<CTransactionRef>& txnConflicted)
+    {
+        BOOST_CHECK_EQUAL(m_expected_tip, block->hashPrevBlock);
+        BOOST_CHECK_EQUAL(m_expected_tip, pindex->pprev->GetBlockHash());
+
+        m_expected_tip = block->GetHash();
+    }
+
+    void BlockDisconnected(const std::shared_ptr<const CBlock>& block)
+    {
+        BOOST_CHECK_EQUAL(m_expected_tip, block->GetHash());
+
+        m_expected_tip = block->hashPrevBlock;
+    }
+};
+
+std::shared_ptr<CBlock> Block(const uint256& prev_hash)
+{
+    static int i = 0;
+    static uint64_t time = Params().GenesisBlock().nTime;
+
+    CScript pubKey;
+    pubKey << i++ << OP_TRUE;
+
+    auto ptemplate = BlockAssembler(Params()).CreateNewBlock(pubKey, false);
+    CBlock* pblock = new CBlock(ptemplate->block);
```

Comment:
I'd think this could be written more simply as:

```
auto block = make_shared<CBlock>(ptemplate->block);
block->hashPrevBlock = ...
return block;
```

It looks like pblock is leaked currently, or is this not the case?